toolkit/tools/imagegen/diskutils/initramfs.go

```diff
@@ -0,0 +1,229 @@
+	inputFile, err := os.Open(initramfsPath)
+	if err != nil {
+		return
+	}
+	defer inputFile.Close()
+
+	gzReader, err := pgzip.NewReader(inputFile)
+	if err != nil {
+		return
+	}
+	defer gzReader.Close()
+	cpioReader := cpio.NewReader(gzReader)
+	//cpio.Reader has no Close() function to defer
+
+	// Caller must use InitramfsMount.Close() to clean up the outputs
+	initramfs.outputBuffer = new(bytes.Buffer)
+	initramfs.pgzWriter = pgzip.NewWriter(initramfs.outputBuffer)
+	initramfs.cpioWriter = cpio.NewWriter(initramfs.pgzWriter)
+
+	var bytesIO int64
+	var nextFileHeader *cpio.Header
+	for {
+		var linkPayload []byte
+
+		nextFileHeader, err = cpioReader.Next()
+		if err == io.EOF {
+			err = nil
+			break
+		}
+		if err != nil {
+			return
+		}
+
+		// For a given symlink, generate a payload that contains the read value of the link.
+		// The payload should be written after the header.
+		// e.g. A link from (/bin -> /usr/bin) would have a payload of "/usr/bin").
+		// cpio.ModeSymlink is two bits (one of which is resused), need to check for actual
+		// equality rather than non-zero after masking
+		isLink := (nextFileHeader.Mode&cpio.ModeType == cpio.ModeSymlink)
+		if isLink {
+			linkPayload = []byte(nextFileHeader.Linkname)
+			nextFileHeader.Size = int64(len(linkPayload))
+		}
+
+		err = initramfs.cpioWriter.WriteHeader(nextFileHeader)
+		if err != nil {
+			return
+		}
+
+		if isLink {
+			var bytesWrittenInt int
+
+			// Write returns an int, cast it to an int64 afterwards
+			logger.Log.Tracef("Creating link %s -> %s", nextFileHeader.Name, nextFileHeader.Linkname)
+			bytesWrittenInt, err = initramfs.cpioWriter.Write(linkPayload)
+			bytesIO = int64(bytesWrittenInt)
+
+		} else {
+			bytesIO, err = io.Copy(initramfs.cpioWriter, cpioReader)
+		}
+
+		if err != nil {
+			return
+		}
+
+		logger.Log.Tracef("File %s caused %d bytes to be transfered to new archive", nextFileHeader.Name, bytesIO)
+		logger.Log.Tracef("Buffer unread length: %d", initramfs.outputBuffer.Len())
+	}
+
+	inputFile.Close()
+
+	fileInfo, err := os.Stat(initramfsPath)
+	if err != nil {
+		return
+	}
+
+	// We can't edit a CPIO archive in place, completely overwrite the file with truncate
+	// The output buffer in memory will be used to re-create the initramfs.
+	initramfs.initramfsOutputFile, err = os.OpenFile(initramfsPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileInfo.Mode())
+
+	return
+}
+
+// Close flushes the archives and closes all initramfs resources
+func (i *InitramfsMount) Close() (err error) {
+	var bytesIO int
+
+	logger.Log.Debugf("Closing initramfs file '%s'", i.initramfsOutputFile.Name())
+
+	// Defer close calls to make sure we handle any errors, failing to
+	// close the file means we can't close the install root.
+	defer i.initramfsOutputFile.Close()
+	defer i.pgzWriter.Close()
+	defer i.cpioWriter.Close()
+
+	err = i.cpioWriter.Close()
+	if err != nil {
+		logger.Log.Errorf("Failed to close initramfs: '%s'", err.Error())
```

Comment:
>Failed to close initramfs [](start = 21, length = 25)

Might be worth to make this error sound differently than the one below. Let's mention "cpio" somewhere here and "pigz" below for instance.